sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIOTest.java

```diff
@@ -581,6 +670,54 @@ public void testReadingWithSubSplits() throws Exception {
     assertSourcesEqualReferenceSource(source, splits, null /* options */);
   }
 
+  /** Tests reading all rows from a sub-split table with several key ranges. */
```

Comment:
Not sure I understand how this test is different from testReadingWithSplitsWithSeveralKeyRanges - what is a "sub-split" table?

Reply:
A sub-split table is one that calls splitKeyRangeIntoBundleSizedSubranges() in BigTableSource. I.e. when splitting, the sample size is larger than the desired bundle size, so not only do we split into samples, but within these sample ranges we split again on desired size. E.g. our sample keys happen every 500 bytes, but we want each source to have a bundle of 100 bytes. So this method would split these 500 byte sample ranges into 5 splits of 100 bytes.

Extending the above example with multiple ranges, say we requested to read the ranges [0, 600) and [600, 1000). this splitKeyRangeInto...() method would then be called to split ranges [0, 500), [500, 600), and [600, 1000]. Each of these ranges would then be split equally in 5 parts, since our sample size (500) is 5x our desired bundle size (100). For the [600, 1000) range this would produce 5 80-byte ranges, which is close enough to the desired 100-byte size. The [500, 600) range is the interesting one, since we'll get 5 20-byte ranges instead of a cleaner single 100-byte range. But I digress; the reason the test is here is to make sure that sub-splitting a range that hit the boundaries of a multiple range request are still sub-split.